Ubuntu Core enablement changes

artiepoole opened the pull request:
In an ongoing attempt to enable fpga device functionality in Ubuntu Core, three issues have arisen in relation to the requirement for snaps to be strictly confined. In our case, we are attempting to redistribute dfx-mgr (and libdfx) as part of [FPGAd](github.com/canonical/fpgad). Once the concept of "softeners" is in place, this snap will enable downstream customers to use dfx-mgr as they expect to be able to, but with the caveat that FPGAd will be the gatekeeper for accessing the FPGA subsystems.

In these efforts, we found three challenges in the way dfx-mgr and libdfx operate:

1. both parts of this application write to a file called `state.txt`. In real operation, this file is created at `/state.txt` which is not allowed in a strictly confined snap, and no layout can be applied to files in the root directory, except as files inside known pre-defined directories. 
2. files are freely moved in and our of /lib/firmware. In a snap's environment, this directory does not exist, so accesses to it fail. Repeatedly copying is bad for performance as well.
3. the device tree overlay subsystem's files (configfs) are remounted to `/configfs`. This mount call requires strong permissions, which the store team can't/wont to grant to the fpgad snap. 

In order to tackle these three problems, I have edited dfx-mgr and libdfx with the following solutions:

1) for dfx-mgr, state.txt is now moved to ~`/etc/dfx-mgrd/state.txt`~ `/run/dfx-mgrd/state.txt`. I would actually prefer to remove the "system(command)" calls to mitigate the need for this file at all, but that is out of scope here. 
2) Before anything which will trigger the kernel to look for firmware, instead of copying files to /lib/firmware, the containing directory is written to `/sys/module/firmware_class/parameters/path` instead. The implementation is in libdfx.c/.h to avoid code duplication.
3) the calls to mount `/configfs` are removed, and all paths pointing to "/configfs/"  are replaced by `/sys/kernel/config/` 

Please see https://github.com/Xilinx/libdfx/pull/6 for the libdfx changes

Please do not hesitate to suggest changes or ask for clarification.

artiepoole committed:
daemon_helper.c: remove (most) system(..) calls

This also removes the need for a `state.txt` file, so we stop
creating the `/run/dfx-mgrd` directory because it is no longer
necessary

Signed-off-by: Artie Poole <stuart.poole@canonical.com>

artiepoole committed:
daemon-helper.c: configfs no longer being remounted, use already existing location

Use `/sys/kernel/config/device-tree/overlays/` instead of remounting

Signed-off-by: Artie Poole <stuart.poole@canonical.com>

artiepoole committed:
daemon_helper.c: stops copying files into /lib/firmware

Note: The implementation for controlling firmware is implemented in
libdfx.h/.c to avoid code duplication.

The kernel provides a mechanism to set where firmware loads will look to
match filenames during probe events.
This file is located at `/sys/module/firmware_class/parameters/path`.
These changes remove the need for copying firmware files into /lib/firmware.
The /lib/firmware location is still probed if the file is not found in the
custom location. See
https://docs.kernel.org/driver-api/firmware/fw_search_path.html for more
information.

Signed-off-by: Artie Poole <stuart.poole@canonical.com>

tjaysukh commented:
@artiepoole  We’ll check this and get back in a week.

artiepoole commented:
@tjaysukh, Thank you for the haste! I created this as a draft. There have been some tweaks since creation, so please take these into account during assessment. I have now tested it more thoroughly and have marked this as "ready for review"

tjaysukh commented:
@artiepoole,
Thank you for suggesting these changes and for the detailed explanation. We appreciate the effort you’ve put into this. Here are my thoughts on your commits:
1.**Move /state.txt to /run/dfx-mgrd/state.txt** -  Since we are already modifying the code, let's remove state.txt altogether and replace the system(command) calls with direct file I/O operations as you suggested.
2.**Configfs no longer being remounted** - I agree with your solution and we can adopt this approach.
3.**Stops copying files into /lib/firmware** - Since this commit is dependent on libdfx (specifically the dfx_set_firmware_lookup_path() function). We can incorporate this change only after the updates are integrated into libdfx. 

Please let me know your thoughts.

artiepoole commented:
@tjaysukh 

Thanks for the feedback. 

On point 1), yes, I am happy to implement the direct I/O. I'll do in a follow-up commit. I will also adopt that for libdfx as well, if the responsible team agrees.

I also agree with point 3) about waiting for libdfx before merging.

I'll mark this PR as draft until the direct IO is implemented and update you here when I think it is ready for review.
Please don't hesitate to suggest any other modifications to my suggested changes.

Best,
Artie

artiepoole commented:
Hey @tjaysukh

Please see the latest commit for changes regarding removing system calls. Please note that I did not touch any parts of the repository which I am not familiar with (e.g. contents of `rpu.c` or the `update_env` function).

Edit: If I am to remove the state.txt from libdfx, several of these functions could be moved to libdfx to reduce duplication. This got me thinking - why are they separate projects - are there other projects also using libdfx? 
Do you agree that these functions should/could be moved to libdfx?
Specifically:
- read_single_line
- write_string_to_file
- fpga_state
- check_overlay_was_applied
- write_path_to_overlay
- write_to_fpga_firmware/write_to_fpga_flags
- remove_overlay_dir
- strip_trailing

talhaHavadar commented:
@artiepoole , I think libdfx should only expose dfx related symbols not the generic utility functions, any further change to ABI of those functions would(should) trigger a library transition for the tools depending libdfx.

Ideally I would like to see those common utility functions sourced somewhere common for both libdfx and dfx-mgr but not tied to the library itself.

talhaHavadar commented on src/daemon_helper.c:
```diff
@@ -60,6 +66,97 @@ not_dir(char *path)
+ * Strips trailing needle from haystack - e.g. `\n` from file read
+ * results or `/` from paths before concatenating.
+ */
+void strip_trailing(char *haystack, const char needle)
```
can we make this static as well?

talhaHavadar commented on src/daemon_helper.c:
```diff
@@ -60,6 +66,97 @@ not_dir(char *path)
+    }
+
+    if (!fgets(buf, (int)size, f)) {
+        DFX_ERR("Failed to read from `%s`", path);
```
maybe using `perror` or `strerror` would give more information to us here.

talhaHavadar commented on src/daemon_helper.c:
```diff
@@ -60,6 +66,97 @@ not_dir(char *path)
+{
+    FILE *f = fopen(path, "w");
+    if (!f) {
+        DFX_ERR("Failed to open `%s` for writing", path);
```
Same perror / strerror suggestion here

talhaHavadar commented on src/daemon_helper.c:
```diff
@@ -60,6 +66,97 @@ not_dir(char *path)
+    }
+
+    if (fclose(f) != 0) {
+        DFX_ERR("Failed to close `%s` after writing", path);
```
Same perror / strerror suggestion here

artiepoole replied:
DFX_ERR does strerror in the macro :)

talhaHavadar commented on src/daemon_helper.c:
```diff
@@ -1652,36 +1749,166 @@ firmware_dir_walk(void)
+ * Return: 0 if the overlay status "applied" and path matches requested_path,
+ *        -1 on error or if either assertion is false
+ */
+static int check_overlay_was_applied(char *overlay_dir, char *requested_path)
```
this function modifies overlay_dir so it is not just a check. can we make it so it doesnt do that and use const in arguments to indicate this.

or at worst document this in function header so people are aware that this will change the arguments sent.

For requested_path, I think we are safe to add const easily.

talhaHavadar commented on src/daemon_helper.c:
```diff
@@ -1652,36 +1749,166 @@ firmware_dir_walk(void)
+    char read_buf[128];
+    const char *state_applied = "applied";
+
+    strip_trailing(overlay_dir, '/');
```
we might not need this at all, it is best to verify behaviour for `//path` though

artiepoole replied:
I will test. usually works for bash cmds otherwise I will copy the buf before modifying

artiepoole replied:
I decided to make a copy into full_path. strip trailing and assemble the full path in 3 steps instead. This functionality is now moving into libdfx, however, so have made those changes there

talhaHavadar commented on src/daemon_helper.c:
```diff
@@ -1652,36 +1749,166 @@ firmware_dir_walk(void)
+ * Return: 0 on success,
+ *        -1 on error (e.g., failed to open, write, or close the file)
+ */
+static int write_path_to_overlay(char *overlay_dir, const char *requested_path)
```
const overlay_dir

talhaHavadar commented on src/daemon_helper.c:
```diff
@@ -1693,17 +1920,28 @@ static int fpga_state(void)
  */
 static int user_load_sysfs(char *bin)
```
can we make this argument const as well?

talhaHavadar commented on src/daemon_helper.c:
```diff
@@ -1716,67 +1954,42 @@ static int user_load_sysfs(char *bin)
-	if (system(command)) {
-		DFX_ERR("Failed system() API");
-	}
+static int user_load_overlay(char *ov, char *region) {
```
can we make arguments const

talhaHavadar commented on src/daemon_helper.c:
```diff
@@ -1802,9 +2015,8 @@ static int user_load_overlay(char *ov, char *region)
  *        -1 on failure or if constraints are violated.
  */
-int user_load(int flag, char *binfile, char *overlay, char *region)
+int user_load(const int flag, char *binfile, char *overlay, char *region)
```
didnt check the whole function but please add const to arguments where possible

artiepoole replied:
Sure thing. I was avoiding editing parts I did not touch, but I can go ahead and do so

talhaHavadar commented on src/daemon_helper.c:
```diff
@@ -1853,16 +2065,13 @@ int user_load(int flag, char *binfile, char *overlay, char *region)
+    if (write_to_fpga_flags(flag & 1)) {
+        DFX_ERR("Failed to set flags");
+        goto ret;
+    }
```
I dont see it errors and jumps to the ret on the original file, it could be missed in the original implementation but just wanted to point out the behaviour change here

artiepoole replied:
Ah thanks!

talhaHavadar commented:
I see DFX_ERR already handles strerror so please ignore my comments about strerror: https://github.com/Xilinx/dfx-mgr/blob/master/src/print.h#L43-L47

artiepoole committed:
daemon_helper.c: use const args for user_load related functions

Signed-off-by: Artie Poole <stuart.poole@canonical.com>

artiepoole committed:
dfx-mgr-client.c: add missing newline on load error print

Signed-off-by: Artie Poole <stuart.poole@canonical.com>

artiepoole commented:
Hey @tjaysukh,

I have removed the need for `state.txt`, but in order to do the same in libdfx, I would have had to duplicate a lot of code. I therefore exposed these functions through dfx-mgr, so this PR now depends even more heavily on libdfx changes. 

Just fyi, the whitespace is a mess in daemon_helper.c  - a mix of spaces, 4-width tabs and8-width tabs, all within the same functions and sometimes the same lines. It was very difficult to make changes and maintain these whitespaces. For this reason, I suggest looking through the changes with "ignore whitespace" on at first, and then enabling whitespace differences after understanding the functional changes. 

It may be worth enforcing a clantidy and clangformat style for consistency. 

p.s. I tacked on Talha's suggest const changes, but only for the functions I touched. I also added on newline to a load error print which has been annoying me for ages. Let me know if you'd prefer those/anything else here separately.

tjaysukh commented:
@artiepoole,

Thanks again for the changes and your efforts. I’ll review and test everything, then get back to you
 
> I have removed the need for state.txt, but in order to do the same in libdfx, I would have had to duplicate a lot of code. I therefore exposed these functions through dfx-mgr, so this PR now depends even more heavily on libdfx changes.

I need to discuss this with the stakeholders. As of now, in the lightweight use-case flow, dfx-mgr is not dependent on libdfx, but with this change, as you mentioned it will become heavily dependent on libdfx. I want to ensure everyone is aligned before we proceed.

artiepoole commented:
Hi @tjaysukh, 

> As of now, in the lightweight use-case flow, dfx-mgr is not dependent on libdfx, but with this change, as you mentioned it will become heavily dependent on libdfx.

If necessary, it would not be difficult to recreate the functionality in dfx-mgr standalone. I was not aware of the need for this mode :) Just let me know what the ultimate decision is.

artiepoole commented:
Hey @tjaysukh, 

I'm just bumping to ask if has there been any update? 
If so, please let me know how to proceed.

Best,
Artie

tjaysukh commented:
@artiepoole,
I’ve confirmed with all stakeholders, and everyone is aligned on this approach. We can proceed with testing and incorporating this change once the libdfx updates are complete.

artiepoole commented:
Hey @tjaysukh , I wanted to update you that the libdfx changes have been merged. Please see [here](https://github.com/Xilinx/libdfx/pull/6#issuecomment-3815434024) for the confirmation.

Please keep me updated on the process

Best,
Artie

tjaysukh commented on src/daemon_helper.c:
```diff
@@ -2020,16 +2066,16 @@ int user_unload(int handle)
+	if (FD)
+		closedir(FD);
+	else {
+		DFX_ERR("/sys/kernel/config/device-tree-overlays/ not present on the"
```
Could you correct the path in error message.
"/sys/kernel/config/device-tree-overlays/" to "sys/kernel/config/device-tree/overlays/"

artiepoole replied:
A great spot, thank you!

I fixed this string in the earliest commit to create this DFX_ERR call, but then addressed this issue more properly by using the `DTBO_ROOT_DIR` global value, in place or this hardcoded string, in the commit which introduced that global variable. I hope this is suitable.

tjaysukh commented on src/daemon_helper.h:
```diff
@@ -44,8 +44,8 @@ char *listAccelerators();
-int user_unload_overlay(char *region);
+int user_load(int flag, const char *binfile, const char *overlay, const char *region);
+int user_unload_overlay(const char *region);
 int user_unload(int handle);
```
Please updated the user_unload() declaration with const as function updated.
int user_unload(const int handle)

artiepoole replied:
I am willing to do this, but it's not typical for values passed by value - this value of `handle` is a local copy inside the function, and so the const here does nothing (as in, it is ignored in the header file) - at least that is what I understand

tjaysukh replied:
What I mean is that the function definition in src/daemon_helper.c was updated to int user_unload(const int handle). but the corresponding function declaration does not match this definition.

artiepoole replied:
Ah, I understood that.  I pushed back for the reasons explained below, but have just learned that it is preference, and so did as you asked :saluting_face: - sorry for any frustration.

my point of view
---
it is explained in the C/C++ standards that it does not matter, and clangtidy regards it as a linter warning/redundant code. In fact, my IDE suggests to remove the `const` from the header declaration. That's why it wasn't there and I didn't want to put it there.

The reference for the redundancy is point 3 of the Parameter-type-list section of [CPP Function specs](https://en.cppreference.com/w/cpp/language/function.html)

tjaysukh commented:
@artiepoole,

Thanks for the update, I am currently testing the changes will keep you posted.

tjaysukh commented:
@artiepoole ,

I have completed testing the dfx-mgr application on my side, and everything looks good. Before we merge it to GitHub, it will go through the internal regression testing. Once that is completed, we will proceed with the merge.

tjaysukh commented:
@artiepoole 

We’ve incorporated these changes into the master branch, Here are the commits:
https://github.com/Xilinx/dfx-mgr/commit/29876c24cb39d1cfde7831e87c4989b5ff05606f
https://github.com/Xilinx/dfx-mgr/commit/6c5e53e4c47bf20ee835c0954e44c7d4efff63b2
https://github.com/Xilinx/dfx-mgr/commit/ffedcc843f1b970f29aad209d601d9e3fbec94b1
https://github.com/Xilinx/dfx-mgr/commit/0e99a4589a4dc99b0746a81d79d6feb3d09d0c78
https://github.com/Xilinx/dfx-mgr/commit/af01d7a70d2a044dcdebbdfad10f3f7be7a787f1

Closing this PR.

Thank you.

artiepoole commented:
Thank you @tjaysukh !
